File: py_m/env_.py
class Environment:
    def __init__(self):
        self.store = {}
        self.outer = None

    # 辞書のデフォルト引数は、違うオブジェクトを作ったとき
    # 引数を省略して関数を呼び出すと、前の値が使われるので注意
    #                           ↓ アウト
    # def __init__(self, store={}, outer=None):
    #     self.store = store
    #     self.outer = outer

    def Get(self, name):
        obj = self.store.get(name)
        if obj is None and self.outer is not None:
            obj = self.outer.Get(name)

        return obj

    def Set(self, name, val):
        self.store[name] = val
        return val

    def __str__(self):
        return "Environment"


def NewEnvironment():
    return Environment()


def NewEnclosedEnvironment(outer):
    e = NewEnvironment()
    e.outer = outer
    return e

File: nim_m/env.nim
import tables
import obj


type Environment* = ref object
  store*: Table[string, obj.Obj]
  outer*: Environment


proc get*(self: Environment, name: string): (obj.Obj, bool) =
  var t: (obj.Obj, bool)
  if self.store.hasKey(name):
    t = (self.store[name], true)
  else:
    t = (nil, false)
    if self.outer != nil:
      t = self.outer.get(name)
  return t


proc set*(self: Environment, name: string, val: obj.Obj): obj.Obj =
  self.store[name] = val
  return val


proc NewEnvironment*(): Environment =
  return Environment(store: initTable[string, obj.Obj]())


proc NewEnclosedEnvironment*(outer: Environment): Environment =
  result = NewEnvironment()
  result.outer = outer

